Meraki organization module

kbreit opened the pull request:
##### SUMMARY
Meraki's web based infrastructure uses organizations as the top level data structure component. This module is able to create, modify, list, and clone organizations using Meraki's REST based API.

Note: Meraki does not allow organization deletion through the API. This means integration tests cannot be created as there is no way to programmatically reset this to be a 'clean slate'. If this is a problem for the approval of this module, please let me know so I can focus my efforts on modules which can have integration tests created.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
meraki

##### ANSIBLE VERSION
```
ansible 2.6.0 (meraki/meraki_organization 591278464b) last updated 2018/04/12 17:56:39 (GMT -500)
  config file = None
  configured module search path = ['/Users/kbreit/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/kbreit/Documents/Programming/ansible/lib/ansible
  executable location = /Users/kbreit/Documents/Programming/ansible/bin/ansible
  python version = 3.5.4 (default, Feb 25 2018, 14:56:02) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.39.2)]
```


##### ADDITIONAL INFORMATION
This pull request includes multiple components related to the Meraki modules including: module, documentation fragment, and module utility. Much more testing and development needs to go into all components.

kbreit committed:
Initial commit

Query an organization within Meraki. No support is in place for managing
or creating yet

kbreit committed:
Change output_level method and make the state parameter required.

kbreit committed:
Implemented listing all organizations

- Updated documentation
- Parse results and return all organizations
- Parse results and return specified organization

kbreit committed:
Framework for creating an organization

- Documentation example for organization creation
- Framework exists for creating organizations, pending PR 36809
- Created functions for HTTP calls
- Renamed from dashboard.meraki.com to api.meraki.com
- Added required_if for state

kbreit committed:
Remove absent state

- Meraki API does not support deleting an organization so absent is removed
- Updated documentation to call it state instead of status

kbreit committed:
Small change to documentation

kbreit committed:
Meraki documentation fragment

- Created initial documentation fragment for Meraki modules

kbreit committed:
Support all parameters associated to organization

- Added all parameters needed for all organization actions.
- None of the added ones work at this time.
- Added documentation for clone.

kbreit committed:
Integration test for meraki_organization module

kbreit committed:
Rename module to meraki for porting to module utility

kbreit committed:
Add meraki module utility to branch. Formerly was on a separate branch.

kbreit committed:
CRU support for Meraki organization module

* CRU is supported for Meraki organizations
* There is no DELETE function for organizations in the API
* This code is very messy and needs cleanup
* Create and Update actions don't show status as updated, must fix

kbreit committed:
Added Meraki module utility to module utility documentation list

kbreit committed:
Added support for organization cloning
* Renamed use_ssl to use_https
* Removed define_method()
* Removed is_org()
* Added is_org_valid() which does all org sanity checks

gundalow commented on test/integration/targets/meraki_organization/tasks/main.yml:
```diff
@@ -0,0 +1,20 @@
+- name: List all organizations
+    meraki_organiztaion:
```
Indentation is wrong here, I believe this should be:
```yaml
- name: List all organizations
  meraki_organiztaion:
    auth_key: abc123
    status: query
  delegate_to: localhost
```
Which is causing the yamllint issue.

gundalow commented on test/integration/targets/meraki_organization/tasks/main.yml:
```diff
@@ -0,0 +1,20 @@
+    name: YourOrg
+    status: present
+  delegate_to: localhost
+  
```
Can remove line 20 (ensuring that 19 ends correctly)

kbreit commented:
Based on the importance of integration tests, do others feel this module is viable considering I won't be able to delete organizations programmatically? If it's not viable, I'd rather move my efforts to other modules.

dagwieers commented:
Sure, it's written and works according to what the Meraki API supports so, yes.

kbreit committed:
Fixes for ansibot
- Changed default of use_proxy from true to false
- Removed some commented out code
- Updated documentation

kbreit committed:
Changes for ansibot
- Removed requirement for state parameter. I may readd this.
- Updated formatting

diff --git a/lib/ansible/module_utils/network/meraki/meraki.py b/lib/ansible/module_utils/network/meraki/meraki.py
index 3acd3d1038..395ac7c4b4 100644
--- a/lib/ansible/module_utils/network/meraki/meraki.py
+++ b/lib/ansible/module_utils/network/meraki/meraki.py
@@ -42,7 +42,7 @@ def meraki_argument_spec():
     return dict(auth_key=dict(type='str', no_log=True, fallback=(env_fallback, ['MERAKI_KEY'])),
                 host=dict(type='str', default='api.meraki.com'),
                 name=dict(type='str'),
-                state=dict(type='str', choices=['present', 'absent', 'query'], required=True),
+                state=dict(type='str', choices=['present', 'absent', 'query']),
                 use_proxy=dict(type='bool', default=False),
                 use_https=dict(type='bool', default=True),
                 validate_certs=dict(type='bool', default=True),
diff --git a/lib/ansible/modules/network/meraki/meraki_organization.py b/lib/ansible/modules/network/meraki/meraki_organization.py
index 923d969366..3789be91d6 100644
--- a/lib/ansible/modules/network/meraki/meraki_organization.py
+++ b/lib/ansible/modules/network/meraki/meraki_organization.py
@@ -20,11 +20,9 @@ short_description: Manage organizations in the Meraki cloud
 version_added: "2.6"
 description:
 - Allows for creation, management, and visibility into organizations within Meraki
-
 notes:
 - More information about the Meraki API can be found at U(https://dashboard.meraki.com/api_docs).
 - Some of the options are likely only used for developers within Meraki
-
 options:
     name:
         description:
@@ -32,21 +30,18 @@ options:
         - If C(clone) is specified, C(name) is the name of the new organization.
     state:
         description:
-        - Create or query organizations
-        choices: ['query', 'present']
+        - Create or modify an organization
+        choices: ['present', 'query']
     clone:
         description:
         - Organization to clone to a new organization.
-        type: string
     org_name:
         description:
         - Name of organization.
         - Used when C(name) should refer to another object.
-        type: string
     org_id:
         description:
         - ID of organization
-
 author:
     - Kevin Breit (@kbreit)
 extends_documentation_fragment: meraki
@@ -86,7 +81,6 @@ RETURN = '''
 response:
     description: Data returned from Meraki dashboard.
     type: dict
-    state: query
     returned: info
 '''

@@ -103,6 +97,7 @@ def main():
     argument_spec = meraki_argument_spec()
     argument_spec.update(clone=dict(type='str'),
+                         state=dict(type='str', choices=['present', 'query']),
                          )

@@ -125,11 +120,9 @@ def main():

     meraki.function = 'organizations'
     meraki.params['follow_redirects'] = 'all'
-    meraki.required_if=[
-                           ['state', 'present', ['name']],
-                           ['clone', ['name']],
-                           # ['vpn_PublicIP', ['name']],
-                       ]
+    meraki.required_if = [['state', 'present', ['name']],
+                          ['clone', ['name']],
+                          ]

     create_urls = {'organizations': '/organizations',
                    }
@@ -162,23 +155,16 @@ def main():

-
-    # method = None
-    # org_id = None
-
-
-    # meraki.fail_json(msg=meraki.is_org_valid(meraki.get_orgs(), org_name='AnsibleTestOrg'))
-
     if meraki.params['state'] == 'query':
-      if meraki.params['name'] is None:  # Query all organizations, no matter what
-        orgs = meraki.get_orgs()
-        meraki.result['organization'] = orgs
-      elif meraki.params['name'] is not None:  # Query by organization name
-        module.warn('All matching organizations will be returned, even if there are duplicate named organizations')
-        orgs = meraki.get_orgs()
-        for o in orgs:
-          if o['name'] == meraki.params['name']:
-            meraki.result['organization'] = o
+        if meraki.params['name'] is None:  # Query all organizations, no matter what
+            orgs = meraki.get_orgs()
+            meraki.result['organization'] = orgs
+        elif meraki.params['name'] is not None:  # Query by organization name
+            module.warn('All matching organizations will be returned, even if there are duplicate named organizations')
+            orgs = meraki.get_orgs()
+            for o in orgs:
+                if o['name'] == meraki.params['name']:
+                    meraki.result['organization'] = o
     elif meraki.params['state'] == 'present':
         if meraki.params['clone'] is not None:  # Cloning
             payload = {'name': meraki.params['name']}
@@ -193,7 +179,10 @@ def main():
             payload = {'name': meraki.params['name'],
                        'id': meraki.params['org_id'],
                        }
-            meraki.result['response'] = json.loads(meraki.request(meraki.construct_path('update', org_id=meraki.params['org_id']), payload=json.dumps(payload), method='PUT'))
+            meraki.result['response'] = json.loads(meraki.request(meraki.construct_path('update',
+                                                                                        org_id=meraki.params['org_id']),
+                                                                  payload=json.dumps(payload),
+                                                                  method='PUT'))

diff --git a/lib/ansible/utils/module_docs_fragments/meraki.py b/lib/ansible/utils/module_docs_fragments/meraki.py
index e268d02e68..3569d83b99 100644
--- a/lib/ansible/utils/module_docs_fragments/meraki.py
+++ b/lib/ansible/utils/module_docs_fragments/meraki.py
@@ -35,6 +35,7 @@ options:
         description:
         - Set amount of debug output during module execution
         choices: ['normal', 'debug']
+        default: 'normal'
     timeout:
         description:
         - Time to timeout for HTTP requests.
diff --git a/test/integration/targets/meraki_organization/aliases b/test/integration/targets/meraki_organization/aliases
new file mode 100644
index 0000000000..ad7ccf7ada
--- /dev/null
+++ b/test/integration/targets/meraki_organization/aliases
@@ -0,0 +1 @@
+unsupported

kbreit committed:
Formatting fix

kbreit committed:
Minor updates due to testing
- Made state required again
- Improved formatting for happier PEP8
- request() now sets instance method

kbreit commented:
Latest commit is ready for review for merge.

kbreit committed:
Fix reporting of the result

kbreit committed:
Enhance idempotency checks
- Remove merging functionality as the proposed should be used
- Do check and reverse check to look for differences

kbreit committed:
Rewrote and added additional integration tests. This isn't done.

kbreit committed:
Updated is_update_required method:
- Original and proposed data is passed to method
- Added ignored_keys list so it can be skipped if needed

dagwieers commented on lib/ansible/module_utils/network/meraki/meraki.py:
```diff
@@ -0,0 +1,370 @@
+def meraki_argument_spec():
+    return dict(auth_key=dict(type='str', no_log=True, fallback=(env_fallback, ['MERAKI_KEY'])),
+                host=dict(type='str', default='api.meraki.com'),
+                name=dict(type='str'),
```
I would use more specific names as part of the module (e.g. org, organization or org_name) but keep an alias for name specific to what the module is managing.

dagwieers commented on lib/ansible/module_utils/network/meraki/meraki.py:
```diff
@@ -0,0 +1,370 @@
+        self.url = None
+
+        self.get_urls = {'organizations': '/organizations',
+                         'networks': '/organizations/replace_org_id/networks',
```
I would do this in a more pythonic way like, `'/organizations/{0}/networks'`

kbreit replied:
I am cleaning it up but the logic isn't working. I'm using the {} method but the logic isn't much different.

kbreit replied:
9df64c0 addresses this comment.

dagwieers commented on lib/ansible/module_utils/network/meraki/meraki.py:
```diff
@@ -0,0 +1,370 @@
+            self.method = method
+        self.url = '{0}://{1}/api/v0/{2}'.format(self.params['protocol'], self.params['host'], self.path.lstrip('/'))
+        if payload is None:
+            resp, info = fetch_url(self.module, self.url,
```
So this is not required, if payload=None it will work fine.

kbreit reviewed (commented):
All changes applied except for simplifying the {} logic in construct_path().

kbreit committed:
Changes per comments from dag
- Optionally assign function on class instantiation
- URLs now have {} for substitution method
- Move auth_key check to module utility
- Remove is_new and get_existing
- Minor changes to documentation

kbreit committed:
Enhancements for future modules and organization
- Rewrote construct_path method for simplicity
- Increased support for network functionality to be committed

dagwieers commented on lib/ansible/module_utils/network/meraki/meraki.py:
```diff
@@ -0,0 +1,408 @@
+
+        if module.params['auth_key'] is None:
+            module.fail_json(msg='Meraki API key not specified')
+        else:
```
There's no need to do the 'else', because everything that follows is also your 'else'. It's the continuation of your program.

dagwieers commented on lib/ansible/module_utils/network/meraki/meraki.py:
```diff
@@ -0,0 +1,408 @@
+            self.module.warn('Enable debug output because ANSIBLE_DEBUG was set or output_level is set to debug.')
+
+        # TODO: This needs to be tested
+        self.module.required_if = [('state', 'present', ['name']),
```
This won't work if 'name' is an alias, rather than the parameter name.

dagwieers commented on lib/ansible/module_utils/network/meraki/meraki.py:
```diff
@@ -0,0 +1,408 @@
+
+        if self.params['auth_key'] is None:
+            try:
+                self.params['auth_key'] = os.environ['MERAKI_KEY']
```
Didn't we already took care of this a little bit up ?

kbreit replied:
We moved it into the module utility.

dagwieers replied:
No, I mean that at line 60 we already do this:
```python
       if module.params['auth_key'] is None:
           module.fail_json(msg='Meraki API key not specified')
       else:
```

And at the very start of the module, we read it from the environment as fallback mechanism. So this is completely useless.

kbreit replied:
Fixed.

dagwieers replied:
I still don't get it. This was taken care of also in the arg_spec. So I would have left line 60 as it was and removed this.

dagwieers commented on lib/ansible/module_utils/network/meraki/meraki.py:
```diff
@@ -0,0 +1,408 @@
+        try:
+            return json.loads(rawoutput)
+        except Exception as e:
+            self.error = dict(code=-1, text="Unable to parse output as JSON, see 'raw' output. {0}".format(e))
```
This was very specific to ACI, I don't see a good reason to repeat this here.

dagwieers commented on lib/ansible/module_utils/network/meraki/meraki.py:
```diff
@@ -0,0 +1,408 @@
+        ''' Return network id from lookup or existing data '''
+        if not data:
+            meraki.fail_json(msg='Must implement lookup')
+        else:
```
No need for the 'else' here either. The 'if' is the exit statement.

dagwieers commented on lib/ansible/module_utils/network/meraki/meraki.py:
```diff
@@ -0,0 +1,408 @@
+        if self.status >= 400:
+            try:
+                self.error['text'] = self.response_json(info['body'])
+                self.error['code'] = info['status']
```
I don't see the value of setting text/code, when 'code' will be also 'status', and 'text' will be error 'msg'.
For ACI you have the HTTP 'status' and HTTP 'response' and you had the APIC 'code' and 'text'.
But for Meraki you don't have that. So I would get rid of all this.

dagwieers commented on lib/ansible/module_utils/network/meraki/meraki.py:
```diff
@@ -0,0 +1,408 @@
+                self.error['code'] = info['status']
+                self.fail_json(msg='Dashboard API error %(code)s: %(text)s' % self.error)
+            except KeyError:
+                self.fail_json(msg='Connection failed for %(url)s. %(msg)s' % info)
```
In fact you could have one big try-except statement, and only a single KeyError exception clause for the connection error.

dagwieers commented on lib/ansible/module_utils/network/meraki/meraki.py:
```diff
@@ -0,0 +1,408 @@
+            except KeyError:
+                self.fail_json(msg='Connection failed for %(url)s. %(msg)s' % info)
+        if self.status >= 201 and self.status <= 299 and method == 'POST':
+            self.result['changed'] = True
```
This seems dangerous. Why would a status 200 PUT, or a status 201-299 POST always induce a change ?

dagwieers commented on lib/ansible/module_utils/network/meraki/meraki.py:
```diff
@@ -0,0 +1,408 @@
+
+    def fail_json(self, msg, **kwargs):
+        # Return error information, if we have it
+        if self.error['code'] is not None and self.error['text'] is not None:
```
Unneeded.

dagwieers commented on lib/ansible/utils/module_docs_fragments/meraki.py:
```diff
@@ -0,0 +1,49 @@
+        description:
+        - Whether to validate HTTP certificates.
+        type: bool
+        default: 'yes'
```
Why is 'org_name' and 'org_id' not in this list if it is part of the default argument_spec ?

kbreit replied:
Can I override the documentation fragment information in the module itself?

dagwieers replied:
Good question, I would hope so.

kbreit replied:
I added to the documentation fragment so we'll see if a module can override fragments.

kbreit committed:
Changes based on Dag feedback and to debug problems

kbreit committed:
Minor fixes for validitation testing

kbreit committed:
Small changes for dag and Ansibot
- Changed how auth_key is processed
- Removed some commented lines
- Updated documentation fragment, but that may get reverted

kbreit committed:
Remove blank line and comment

dagwieers commented on lib/ansible/module_utils/network/meraki/meraki.py:
```diff
@@ -0,0 +1,353 @@
+        self.function = function
+
+        # error output
+        self.error = dict(code=None, text=None)
```
Unneeded.

kbreit replied:
Dead.

dagwieers commented on lib/ansible/module_utils/network/meraki/meraki.py:
```diff
@@ -0,0 +1,353 @@
+                    }
+        net_only = {'net_id': net_id,
+                    }
+        org_net = {'org_id': org_id,
```
Not sure why you need all this complexity ? (And the complexity below)
I would have expected this to be much more simple.

kbreit replied:
A rewritten version will be in the next commit.

dagwieers commented on lib/ansible/module_utils/network/meraki/meraki.py:
```diff
@@ -0,0 +1,353 @@
+
+        if method is not None:
+            self.method = method
+        self.url = '{0}://{1}/api/v0/{2}'.format(self.params['protocol'], self.params['host'], self.path.lstrip('/'))
```
Did you know you can do this:
```python
self.url = '{protocol}://{host}/api/v0/{path}'.format(path=self.path.lstrip('/'), **self.params)
```

kbreit replied:
I didn't do it this exact way, but the construct_path has been simplified.

dagwieers commented on lib/ansible/module_utils/network/meraki/meraki.py:
```diff
@@ -0,0 +1,353 @@
+                # self.error['code'] = info['status']
+                self.fail_json(msg='Dashboard API error %(code)s: %(text)s' % self.error)
+        except KeyError:
+            self.fail_json(msg='Connection failed for %(url)s. %(msg)s' % info)
```
Looking at the code now, it's wrong. You don't have `self.error` anymore.
And the KeyError will never hit, so you surely didn't test these failures. (integration tests !)

There's also no difference between `>= 300` and `>= 400` so I guess `>= 300` only would do (but I don't know if that is intentional).

kbreit replied:
I believe resolved, but integration tests will need to be written to prove it out.

kbreit committed:
Improvements for testing and code simplification
- Added network integration tests
- Modified error handling in request()
	- More testing to come on this
- Rewrote construct_path again. Very simple now.

kbreit committed:
Remove trailing whitespace

dagwieers commented on lib/ansible/module_utils/network/meraki/meraki.py:
```diff
@@ -0,0 +1,287 @@
+               'net_id': net_id,
+               }
+
+        built_path = built_path.format(**ids)
```
As I indicated yesterday, here you could be doing:
```python
    built_path = build_path.format(org_id=org_id, net_id=net_id)
```
Rather than making a dict first, and then unfolding it for format().
For larger sets of variables you code is probably preferred.

dagwieers commented on lib/ansible/module_utils/network/meraki/meraki.py:
```diff
@@ -0,0 +1,287 @@
+
+        if method is not None:
+            self.method = method
+        self.url = '{0}://{1}/api/v0/{2}'.format(self.params['protocol'], self.params['host'], self.path.lstrip('/'))
```
I suggested yesterday something more convenient and readable:
```python
self.url = '{protocol}://{host}/api/v0/{path}'.format(path=self.path.lstrip('/'), **self.params)
```

dagwieers reviewed (commented):
Except for 2 suggestions, everything is looking good to me. The tests should be improved, but that shouldn't be holding us back to merge this.

kbreit committed:
Small changes based on dag's response

kbreit committed:
Removed certain sections from exit_json and fail_json as they're old

dagwieers reviewed (approved):
LGTM
